### DISCUSSION TITLE
Depuis la version 3.8.0 librairie <SD.H> SD.end(); bloque !

### DISCUSSION THREAD

**Discussion by SatreDaniel** [earlephilhower/arduino-pico#2220] [Q&A]
Depuis la version 3.8.0 librairie <SD.H> SD.end(); bloque ! 
Si je compil avec la version 3.7.2 tout fonction bien
````
  SPI.setRX(4);     // MISO (pin 6)
  SPI.setCS(5);     // CS   (pin 7)
  SPI.setSCK(2);    // SCK  (pin 4)
  SPI.setTX(3);     // MOSI (pin 5)

  if (!tst_SDcard()) {               
    Serial.println("initialization failed !");
  }

 bool tst_SDcard (){                      // Initialise SD Card, créer fichier "datalog.txt"

  int i = 0;
  char  lect = LF;  
  bool ret = true;

  Serial.println("");

  SD.end();
  
  if ( !SD.begin(SDcspin) ) {
    Serial.println("Carte SD retirée ou non initialisée.");
    ret = false;

  } else {
    Serial.println("initialisation faite ...!");              // Lecteur carte a été initialisé
    dataFile = SD.open("/datalog.txt", FILE_READ);
    if (dataFile) {                                           // le Fichier Existe
      Serial.println("File datalog.txt open to read ...");
      for (i=0; i < szData ; i++){                            // On pose une limite (SD a des comportement ératique parfois !)
        if (dataFile.available()) {                             
          lect=dataFile.read();                               // On lit une ligne (pour être sûr !)
          Serial.write(lect);
          if (lect == LF){break;}
        }
      }

    } else {
      dataFile.close();
      dataFile = SD.open("/datalog.txt", FILE_WRITE);           // Création du fichier (Ouverture en écriture)
      if (!dataFile) {
        dataFile.close();
        Serial.println("Error Open or Create datalog.txt..."); 
        ret = false;
      }
    }
    dataFile.close();
  }

  Serial.println("");
  return(ret);
 }
````

**Comment by earlephilhower** [comment]
#2221 fixes.  Issue was calling SD.end() before any SD.begin().

**Reply by SatreDaniel** [reply]
Hello, thank.
But before or after it's the same problem,
And in this example I have a SD.begin() in the setup() and or I put SD.end() after.
I had not problems with "Raspberry Pi Pico/RP2040 3.7.2" 
Best regard

// Configure pins for SPI comms to SD card when used with Raspberry Pi Pico W.
  SPI.setRX(4);     // MISO (pin 6)
  SPI.setCS(5);     // CS   (pin 7)
  SPI.setSCK(2);    // SCK  (pin 4)
  SPI.setTX(3);     // MOSI (pin 5)

  Serial.println("");  
  if ( !SD.begin(SDcspin) ) {
    Serial.println("initialization failed !");
  }
  Serial.println(""); 

//------------

bool tst_SDcard (){                      // Initialise SD Card, créer fichier "datalog.txt"

  int i = 0;
  char  lect = LF;  
  bool ret = true;

  Serial.println("");
  
  if ( !SD.begin(SDcspin) ) {
    Serial.println("Carte SD retirée ou non initialisée.");
    ret = false;

  } else {
    Serial.println("initialisation faite ...!");              // Lecteur carte a été initialisé
    dataFile = SD.open("/datalog.txt", FILE_READ);
    if (dataFile) {                                           // le Fichier Existe
      Serial.println("File datalog.txt open to read ...");
      for (i=0; i < szData ; i++){                            // On pose une limite (SD a des comportement ératique parfois !)
        if (dataFile.available()) {                             
          lect=dataFile.read();                               // On lit une ligne (pour être sûr !)
          Serial.write(lect);
          if (lect == LF){break;}
        }
      }

    } else {
      dataFile.close();
      dataFile = SD.open("/datalog.txt", FILE_WRITE);           // Création du fichier (Ouverture en écriture)
      if (!dataFile) {
        dataFile.close();
        Serial.println("Error Open or Create datalog.txt..."); 
        ret = false;
      }
    }
    dataFile.close();
  }

  Serial.println("");
  SD.end();                                          // On désinitialise le Lecteur
  return(ret);
 }

**Reply by earlephilhower** [reply]
Please post a complete compilable sample that fails with the 1-line change (note, the change was NOT published yet, you need to manually apply it or use a `git clone` of the repo to get it).  I took the snippet you had and fixed it up so it compiled and it was obviously just the `_spi->end()` call seen before `_spi` was set causing a crash.  Once that was fixed, it worked fine and reported `initialision faite!` and all was well...
